### PR TITLE
Add rep delay

### PR DESCRIPTION
**Pull request by zachschoenfeld33**
### Summary
This PR adds a flag `rep_delay` to the Pulse Qobj Schema. `rep_delay` specifies the delay between program executions. A flag `dynamic_reprate_enabled` is retrieved from the backend configuration which indicates whether a `rep_delay` can be used on that backend. If it is false, we default back to the standard behavior of `rep_time`, which defines the duration of a program execution.

For certain experiments, it is advantageous to be able to define the delay between shots rather than the total length of a circuit (since in the latter case there can be overlap in the acquisition stage).

In general, the preferred flag moving forward is `rep_delay` and warnings are thrown to indicate that...

Schemas have been updated with these changes.
### Details and comments
- Add `rep_delay` flag specifying time delay between program executions
- Add backend configuration flag `dynamic_reprate_enabled` which defines whether the dynamic `rep_delay` can be used on a given backend
- Throw a `RuntimeWarning` if `rep_delay` is used and dynamic rep rates are disabled
- Throw a `RuntimeWarning` if `rep_time` is used and dynamic rep rates are enabled
- Assembler tests for these features

### ToDo
- Release notes/Change log?

### CODE REVIEW

**Comment by zachschoenfeld33** [comment]
@taalexander @lcapelluto, this PR adds `rep_delay` in Qiskit Pulse.

File: qiskit/compiler/assemble.py [hunk 1]
```diff
@@ -103,9 +105,12 @@ def assemble(experiments: Union[QuantumCircuit, List[QuantumCircuit], Schedule,
...
         memory_slot_size: Size of each memory slot if the output is Level 0.
-        rep_time: Repetition time of the experiment in s.
-            The delay between experiments will be ``rep_time``.
-            Must be from the list provided by the device.
+        rep_time: Time per program execution in sec. Must be from the list provided
+            by the backend.
+        rep_delay: Delay between programs in sec. Only supported on certain
+            backends (``backend.configuration()['dynamic_reprate_enabled']`` ).
+            If supported, it will override ``rep_time``. Must be from the list
+            provided by the backends.
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:113]: Should share here what list you mean

File: qiskit/compiler/assemble.py [hunk 2]
```diff
@@ -276,13 +283,29 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
...
-    rep_time = rep_time or getattr(backend_config, 'rep_times', None)
 
-    if isinstance(rep_time, list):
-        rep_time = rep_time[0]
+    dynamic_reprate_enabled = getattr(backend_config, 'dynamic_reprate_enabled', False)
 
+    rep_time = rep_time or getattr(backend_config, 'rep_times', None)
     if rep_time:
-        rep_time = int(rep_time * 1e6)
+        # deprecation warning for ``rep_time`` when dynamic rep rate is enabled
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:291]:
```suggestion
```
this comment restates the code

File: qiskit/compiler/assemble.py [hunk 3]
```diff
@@ -276,13 +283,29 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
...
+        if dynamic_reprate_enabled:
+            warnings.warn("Use 'rep_delay' rather than 'rep_time' as dynamic rep rates are enabled "
+                          "on this backend.", DeprecationWarning)
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:294]: How will this work? "Deprecation" means we're going to remove functionality at some point. Will there be a point that `rep_time` can be removed completely, i.e. there won't be any backends that require that rep time is supported?

File: qiskit/compiler/assemble.py [hunk 4]
```diff
@@ -276,13 +283,29 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
...
+        if isinstance(rep_time, list):
+            rep_time = rep_time[0]
+        rep_time = rep_time * 1e6 # convert sec to μs
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:297]: Why did you remove `int`?

File: qiskit/compiler/assemble.py [hunk 5]
```diff
@@ -276,13 +283,29 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
...
+
+    rep_delay = rep_delay or getattr(backend_config, 'rep_delays', None)
+    # if ``rep_delay`` is None, leave as None; otherwise update appropriately
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:300]:
```suggestion
```

File: qiskit/compiler/assemble.py [hunk 6]
```diff
@@ -297,6 +320,7 @@ def _parse_pulse_args(backend, qubit_lo_freq, meas_lo_freq, qubit_lo_range,
                            meas_map=meas_map,
                            memory_slot_size=memory_slot_size,
                            rep_time=rep_time,
+                           rep_delay=rep_delay,
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:323]: What if both are passed? Then the behavior is determined by `dynamic_reprate_enabled`, right? Should only `rep_time` be passed if dynamic_reprate_enabled is False, and only `rep_delay` passed if dynamic_reprate_enabled is True? Maybe it doesn't really matter

File: qiskit/compiler/assemble.py (hunk 4 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py:297]: Well, to my understanding at least, I believe these are floats?

File: qiskit/compiler/assemble.py (hunk 6 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py:323]: Yes. I think it's fine to pass it--it won't influence the behavior...

File: qiskit/compiler/assemble.py (hunk 1 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/d192e28937f3e8826260746c7d083e272ffa2075.

File: qiskit/compiler/assemble.py (hunk 2 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/d192e28937f3e8826260746c7d083e272ffa2075.

File: qiskit/compiler/assemble.py (hunk 3 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py]: Good pt, removed that. Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/d192e28937f3e8826260746c7d083e272ffa2075.

File: qiskit/compiler/assemble.py (hunk 5 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/d192e28937f3e8826260746c7d083e272ffa2075.

File: qiskit/compiler/assemble.py (hunk 4 above)
Comment (lcapelluto) [reply qiskit/compiler/assemble.py:296]: I mean it makes sense to me. Not sure if there was a good reason for it being int before.

File: qiskit/compiler/assemble.py [hunk 7]
```diff
@@ -103,9 +105,12 @@ def assemble(experiments: Union[QuantumCircuit, List[QuantumCircuit], Schedule,
                 * ``avg`` returns average measurement output (averaged over number of shots).
         meas_map: List of lists, containing qubits that must be measured together.
         memory_slot_size: Size of each memory slot if the output is Level 0.
-        rep_time: Repetition time of the experiment in s.
-            The delay between experiments will be ``rep_time``.
-            Must be from the list provided by the device.
+        rep_time: Time per program execution in sec. Must be from the list provided
+            by the backend (``backend.configuration().rep_times``).
+        rep_delay: Delay between programs in sec. Only supported on certain
+            backends (``backend.configuration()['dynamic_reprate_enabled']`` ).
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:111]:
```suggestion
            backends (``backend.configuration().dynamic_reprate_enabled`` ).
```

Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/28b45ffa8eea665a684ffc0ba3d509740788c17f.

File: qiskit/compiler/assemble.py [hunk 8]
```diff
@@ -103,9 +105,12 @@ def assemble(experiments: Union[QuantumCircuit, List[QuantumCircuit], Schedule,
                 * ``avg`` returns average measurement output (averaged over number of shots).
         meas_map: List of lists, containing qubits that must be measured together.
         memory_slot_size: Size of each memory slot if the output is Level 0.
-        rep_time: Repetition time of the experiment in s.
-            The delay between experiments will be ``rep_time``.
-            Must be from the list provided by the device.
+        rep_time: Time per program execution in sec. Must be from the list provided
+            by the backend (``backend.configuration().rep_times``).
+        rep_delay: Delay between programs in sec. Only supported on certain
+            backends (``backend.configuration().dynamic_reprate_enabled`` ).
```
Comment (lcapelluto) [inline qiskit/compiler/assemble.py:111]:
```suggestion
            backends (``backend.configuration().dynamic_reprate_enabled``).
```

File: qiskit/qobj/pulse_qobj.py [hunk 9]
```diff
@@ -249,7 +249,12 @@ def __init__(self, meas_level, meas_return, pulse_library,
                 measurement driver LO's in GHz.
             memory_slot_size (int): Size of each memory slot if the output is
                 Level 0.
-            rep_time (int): Repetition time of the experiment in μs
+            rep_time: Time per program execution in sec. Must be from the list provided
+                by the backend (``backend.configuration().rep_times``).
+            rep_delay: Delay between programs in sec. Only supported on certain
+                backends (``backend.configuration()['dynamic_reprate_enabled']`` ).
```
Comment (lcapelluto) [inline qiskit/qobj/pulse_qobj.py:255]:
```suggestion
                backends (``backend.configuration().dynamic_reprate_enabled``).
```

File: qiskit/providers/models/backendconfiguration.py [hunk 10]
```diff
@@ -477,7 +484,13 @@ def __init__(self,
         self.discriminators = discriminators
         self.hamiltonian = hamiltonian
 
-        self.rep_times = [_rt * 1e-6 for _rt in rep_times]
+        self.dynamic_reprate_enabled = dynamic_reprate_enabled
+
+        self.rep_times = [_rt * 1e-6 for _rt in rep_times]  # convert to sec
```
Comment (lcapelluto) [inline qiskit/providers/models/backendconfiguration.py:489]: Will `rep_times` potentially be `None`?

File: qiskit/compiler/assemble.py (hunk 8 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assemble.py:111]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/bcfd9c4d65c352c231fb033a4239dfdf9f8a2bc4

File: qiskit/qobj/pulse_qobj.py (hunk 9 above)
Comment (zachschoenfeld33) [reply qiskit/qobj/pulse_qobj.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4539/commits/bcfd9c4d65c352c231fb033a4239dfdf9f8a2bc4.

File: qiskit/schemas/qobj_schema.json [hunk 11]
```diff
@@ -1085,8 +1085,14 @@
                             "type": "array"
                         },
                         "rep_time": {
-                            "minimum": 1,
```
Comment (taalexander) [inline qiskit/schemas/qobj_schema.json:1088]: This may be the issue in #4618.
